### PR TITLE
Cleanup with initcalls

### PR DESCRIPTION
**Pull request by jenswi-linaro**

### CODE REVIEW

File: core/include/initcall.h [hunk 1]
```diff
@@ -8,15 +8,30 @@
...
+#if TRACE_LEVEL >= TRACE_DEBUG
+#define __define_initcall(lvl, fn) \
+	SCATTERED_ARRAY_DEFINE_PG_ITEM_ORDERED(initcall, lvl, \
+					       struct initcall) = \
+		{ .func = (fn), .level = (lvl), .func_name = #fn, }
+#else
+#define __define_initcall(lvl, fn) \
+	SCATTERED_ARRAY_DEFINE_PG_ITEM_ORDERED(initcall, lvl, \
+					       struct initcall) = \
+		{ .func = (fn), }
```
Comment (etienne-lms) [inline core/include/initcall.h:30]: missing `.level = (lvl),`

File: core/arch/arm/kernel/tee_time.c [hunk 2]
```diff
@@ -3,18 +3,27 @@
  * Copyright (c) 2016, Linaro Limied
```
Comment (etienne-lms) [inline core/arch/arm/kernel/tee_time.c:3]: 2016-2020 ?

File: core/mm/fobj.c [hunk 3]
```diff
@@ -52,7 +59,9 @@ void fobj_generate_authenc_key(void)
 				      sizeof(rwp_ae_key.data),
 				      &rwp_ae_key.rounds))
 		panic("failed to expand key");
+	return TEE_SUCCESS;
```
Comment (etienne-lms) [inline core/mm/fobj.c:62]: maybe an empty line above.

File: core/arch/arm/kernel/sub.mk [hunk 4]
```diff
@@ -48,6 +48,7 @@ ifeq ($(CFG_GENERIC_BOOT),y)
 srcs-$(CFG_ARM32_core) += generic_entry_a32.S
 srcs-$(CFG_ARM64_core) += generic_entry_a64.S
 endif
+srcs-y += initcall.c
```
Comment (etienne-lms) [inline core/arch/arm/kernel/sub.mk:51]: **initcall.c** is not arm specific? Move to **core/kernel/**?

Comment (jenswi-linaro) [reply core/arch/arm/kernel/sub.mk:51]: True, I'll move it.

File: core/include/initcall.h (hunk 1 above)
Comment (jenswi-linaro) [reply core/include/initcall.h:30]: No, `.level` doesn't exist in the part of the #ifdef

File: core/arch/arm/kernel/tee_time.c (hunk 2 above)
Comment (jenswi-linaro) [reply core/arch/arm/kernel/tee_time.c:3]: I'll fix

File: core/mm/fobj.c (hunk 3 above)
Comment (jenswi-linaro) [reply core/mm/fobj.c:62]: OK

File: core/include/initcall.h (hunk 1 above)
Comment (etienne-lms) [reply core/include/initcall.h:30]: sorry, you're right.

**Comment by jenswi-linaro** [comment]
Comments addressed, tags applied. Note that "core: call call_initcalls() directly" and "￼core: add call_finalcalls()" are missing tags.

**Comment by jforissier** [comment]
For ""core: call call_initcalls() directly"":
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

File: core/arch/arm/kernel/generic_boot.c [hunk 5]
```diff
@@ -1179,7 +1178,6 @@ void __weak paged_init_primary(unsigned long fdt)
 #ifndef CFG_VIRTUALIZATION
 	init_tee_runtime();
 #endif
-	release_external_dt();
```
Comment (jforissier) [inline core/arch/arm/kernel/generic_boot.c:1183]: This changes the order, prior to this patch `release_external_dt()` was called before `core_mmu_init_virtualization()`, now it called after (as part of `call_finalcalls()`). Is it OK?

Comment (etienne-lms) [reply core/arch/arm/kernel/generic_boot.c:1183]: As far as `release_external_dt()` shall be called before non-secure world in entered, it looks good to me.

Comment (etienne-lms) [reply core/arch/arm/kernel/generic_boot.c:1183]: But @lorc may have some comment about it...

Comment (jenswi-linaro) [reply core/arch/arm/kernel/generic_boot.c:1183]: It doesn't seem they have much to do with each other. @lorc?

**Comment by jenswi-linaro** [comment]
Squashed and tags applied.

**Comment by jenswi-linaro** [comment]
Rebased and fixed trivial conflicts

**Pull request merged**